pkg/frontend/v2/frontend.go

```diff
@@ -308,14 +310,30 @@ func (f *Frontend) CheckReady(_ context.Context) error {
 type requestsInProgress struct {
 	mu       sync.Mutex
 	requests map[uint64]*frontendRequest
+	logger   log.Logger
+	wg       sync.WaitGroup
 }
 
-func newRequestsInProgress() *requestsInProgress {
+func newRequestsInProgress(logger log.Logger) *requestsInProgress {
 	return &requestsInProgress{
 		requests: map[uint64]*frontendRequest{},
+		logger:   logger,
 	}
 }
 
+// Wait waits for any in-flight requests to finish.
+func (r *requestsInProgress) Wait() {
+	if r.count() == 0 {
+		level.Info(r.logger).Log("msg", "ready to terminate, no in-flight queries remain")
+		return
+	}
+
+	level.Info(r.logger).Log("msg", "waiting for in-flight queries to finish...")
```

Comment:
I think this log should be moved in the caller.

Reply:
:eyes:

Reply:
Done.